### PR TITLE
Ceciila/opt/pulloutput

### PR DESCRIPTION
**Pull request by ceciliapeng2011**
### Details:
 - *change the default case resolution policy so that it resolves cyclic inplace to downstream instead of upstream for outputs.*
 - *this is a more optimal decision as more cases are going to reuse inferrequest's memory in PullOutputData.*

### Tickets:
 - *116823*

### CODE REVIEW

**Comment by yuxu42** [comment]
@maxnick @dmitry-gorokhov could you please help take a review? Thanks!

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "shared_test_classes/base/ov_subgraph.hpp"
+#include "ngraph_functions/utils/ngraph_helpers.hpp"
+#include "ngraph_functions/builders.hpp"
+
+/*
+The main purpose of the tests is to refactor cyclic inplace resolution in order to make output edge be referenced whenever possible.
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp:10]:
```suggestion
The main purpose of the tests is to test cyclic inplace resolution in order to make sure that output edges are referenced whenever possible.
```

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                        |
+                      Softmax
+                     /     \
+                    /       \
+                   Add     Reshape0
+                    |         |
+                    |         |
+                 Result0   Result1
+
+expect edge Reshape1->Result1 to be referenced by its upstreams, instead of referencing to its upstreams.
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp:34]: Where is `Reshape1` on the scheme?

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        auto result_0 = std::make_shared<ngraph::opset3::Result>(add_0);    // dummy output
+
+        auto reshape_param_0 = ngraph::builder::makeConstant<int>(ov::element::i32, {1}, {0});
+        auto reshape_0 = std::make_shared<ngraph::opset1::Unsqueeze>(soft_max, reshape_param_0);
+        auto result_1 = std::make_shared<ngraph::opset3::Result>(reshape_0);    // target output
+
+        ngraph::ResultVector results = {result_0, result_1};
+        function = std::make_shared<ov::Model>(results, params, "Subgraph1");
+
+        ov::pass::Serialize serializer("Subgraph1.xml", "Subgraph1.bin");
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp:61]: Please remove graph serialization

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+        auto reshape_param_1 = ngraph::builder::makeConstant<int>(ov::element::i32, {1}, {0});
+        auto reshape_1 = std::make_shared<ngraph::opset1::Unsqueeze>(reshape_0, reshape_param_1);
+        auto result_1 = std::make_shared<ngraph::opset3::Result>(reshape_1);    // target output
+
+        ngraph::ResultVector results = {result_0, result_1, result_2};
+        function = std::make_shared<ov::Model>(results, params, "Subgraph2");
+
+        ov::pass::Serialize serializer("Subgraph2.xml", "Subgraph2.bin");
+        serializer.run_on_model(function);
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp:142]: Please remove serialization

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+        auto reshape_param_0 = ngraph::builder::makeConstant<int>(ov::element::i32, {1}, {0});
+        auto reshape_0 = std::make_shared<ngraph::opset1::Unsqueeze>(params.front(), reshape_param_0);
+        auto result_0 = std::make_shared<ngraph::opset3::Result>(reshape_0);    // target output
+
+        ngraph::ResultVector results = {result_0};
+        function = std::make_shared<ov::Model>(results, params, "Subgraph3");
+
+        ov::pass::Serialize serializer("Subgraph3.xml", "Subgraph3.bin");
+        serializer.run_on_model(function);
```
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp:201]: Serialization

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp
Comment (maxnick) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/inplace_resolve_io.cpp:1]: A general comment.
These tests are more or less ok as smoke tests. But, they don't check the most essential part, i.e. the referencing relationships. If the feature is broken somehow (the referencing chain gets different), those tests won't detect the regression. Of course, on the OV API level it's nearly impossible to perform the necessary checks.
But, I'm just wondering if it's possible to write the corresponding unit tests.
Technically, all we need is to get access to the internal cpu graph somehow, read the memory object pointers values and compare them.

**Pull request closed without merging**